src/framework/audio/internal/sequencer.cpp

```diff
@@ -0,0 +1,291 @@
+{
+    std::lock_guard guard(m_syncMutex);
+    midi_track_t track = midiTrack(id);
+    if (!track) {
+        track = createMIDITrack(id);
+    }
+    track->loadMIDI(stream);
+}
+
+void Sequencer::setAudioTrack(ISequencer::track_id id, const std::shared_ptr<audio::IAudioStream>& stream)
+{
+    std::lock_guard guard(m_syncMutex);
+    audio_track_t track = audioTrack(id);
+    if (!track) {
+        track = createAudioTrack(id);
+    }
+    track->load(stream);
+}
+
+mu::async::Channel<ISequencer::audio_track_t> Sequencer::audioTrackAdded() const
+{
+    return m_audioTrackAdded;
+}
+
+void Sequencer::initAudioTrack(ISequencer::track_id id)
+{
+    createAudioTrack(id);
+}
+
+mu::async::Channel<mu::midi::tick_t> Sequencer::midiTickPlayed(ISequencer::track_id id) const
+{
+    if (midi_track_t track = midiTrack(id)) {
+        return track->tickPlayed();
+    }
+    return {};
+}
+
+mu::async::Notification Sequencer::positionChanged() const
+{
+    return m_positionChanged;
+}
+
+float Sequencer::playbackPosition() const
+{
+    return clock()->timeInSeconds();
+}
+
+void Sequencer::setLoop(unsigned int from, unsigned int to)
+{
+    std::lock_guard guard(m_syncMutex);
+    m_loopStart = from;
+    m_loopEnd = to;
+}
+
+void Sequencer::unsetLoop()
+{
+    std::lock_guard guard(m_syncMutex);
+    m_loopStart.reset();
+    m_loopEnd.reset();
+}
+
+std::shared_ptr<Clock> Sequencer::clock() const
+{
+    return m_clock;
+}
+
+void Sequencer::timeUpdate()
+{
+    std::lock_guard guard(m_syncMutex);
+    if (m_loopStart.has_value() && m_loopEnd.has_value()) {
+        if (m_clock->timeInMiliSeconds() >= m_loopEnd) {
+            seek(m_loopStart.value_or(0));
+        }
+    }
+
+    bool willcontinue = false;
+    for (auto& track : m_tracks) {
+        track.second->forwardTime(m_clock->timeInMiliSeconds());
+        willcontinue |= track.second->isRunning();
+    }
+    m_positionChanged.notify();
+    if (!willcontinue) {
+        stop();
+    }
+}
+
+void Sequencer::beforeTimeUpdate(Clock::time_t miliseconds)
+{
+    std::lock_guard guard(m_syncMutex);
+    if (m_nextStatus != m_status) {
+        setStatus(m_nextStatus);
+    }
+
+    if (m_nextSeek.has_value()) {
+        auto miliseconds = m_nextSeek.value_or(0);
+        m_nextSeek.reset();
+
+        m_clock->seekMiliseconds(miliseconds);
+        for (auto& track : m_tracks) {
+            track.second->seek(miliseconds);
+        }
+        m_positionChanged.notify();
+    }
+
+    std::for_each(m_backgroudPlayers.begin(), m_backgroudPlayers.end(), [&miliseconds](auto& player) {
+        player.first += miliseconds;
+        player.second->forwardTime(player.first);
+    });
+
+    m_backgroudPlayers.remove_if([](auto player) {
+        return player.second->status() == IPlayer::Status::Stoped;
+    });
+}
+
+void Sequencer::setStatus(Status status)
+{
+    m_status = status;
+    m_statusChanged.send(m_status);
```

Comment:
The function of an automatic stop at the end of playback has been reduced, for example, we have played all the notation, so there is no sound, and the play button is still in the "playing" state.

Reply:
hmm, looking for code, it should work but doesn't work

Reply:
I found out that the problem is not in the audio engine, but in changing the format for transmitting midi events (now tick events are not transmitted)